CODING_STANDARDS.md

```diff
@@ -263,6 +263,25 @@ rewritten to comply with these rules.
    codebase but should not be refactored, unless larger refactoring around that
    code is taking place.
 
+## C Headers
+
+1. Include only the headers (`*.h`) that are really needed.  Consider
+   using [iwyu](https://include-what-you-use.org/).
+
+1. Use forward declarations to eliminate `#include` directives if
+   possible.
```

Comment:
The purpose of includes is too avoid code duplication. This statement says it's better to duplicate code.
May be you had a bit different thoughts, but in current state this looks improper.

Reply:
It's better to write a single line of forward declaration instead of including a whole header (which in turn includes a dozen other headers, each of which includes another dozen of more headers).
This doesn't duplicate code any more than an `#include` duplicates code.
It's the same amount of code (one line), but the forward declaration is better because it's trivial to process by the compiler, expresses exactly what it does, comes without namespace pollution, without I/O, without parsing more code, and so on.

Reply:
> It's better to write a single line of forward declaration instead of including a whole header

and then fix this forward declaration in all places instead of a single place...

Reply:
What needs to be fixed, in your opinion?

Reply:
If these updates can be automated, and apparently [that is possible](https://include-what-you-use.org/) (never used that tool, though), it might not be that bad. :)

Actually, I was in favor of this include style, if at least system headers also adhered to it – but obviously, they don't (at least most of them).

Reply:
I see what you mean, and I should clarify that only structs and unions (and their typedefs) should be forward-declared, not functions - function prototypes (which are another kind of forward declaration) should not, because prototypes are too complex and indeed too much in motion.
If something gets switched between union and struct (which I think is very rare), then you'll get compiler errors which are trivial to fix (a simple `sed` will fix it) - just like you'd get compiler errors for all callers if you change a function prototype. The latter (which is not affected by my proposal) is not trivial to fix, but still happens all the time and you manage it easily. 
Means: there's no reason to believe that struct/union switches are too difficult to do just because there are forward declarations.

Reply:
> if at least system headers also adhered to it – but obviously, they don't

Many projects like glibc, GCC (libstdc++) and Boost are aware of the problem, and each new version improves - which, of course, causes breakages in lots of buggy software that relies on header bloat in system headers and forgets to include necessary headers. For example, several of my projects received bug reports from people trying to compile my code with GCC13....

But anyway, even if system headers are too bloaty, that's just one more reason why doing it correctly in one's own software. Fewer includes means fewer indirect includes, which means more verification for correctness.

My proposal is really common sense. I'm really surprised about the resistance that suddenly emerges here.